### ISSUE TITLE
One-to-many projection creates many subqueries

### ISSUE THREAD

**Issue by brunolau** [linq2db/linq2db.EntityFrameworkCore#17]
First of all - this tool practically saved my life when migrating from EF6 to EF Core, great job, everything works better than EF Core, instead of one thing

```
LinqToDB.Common.Configuration.Linq.AllowMultipleQuery = true;

var querySample = db.TicketNumbers.Where(p => p.Ticket.Name == "Lacna").Select(p => new
{
    Id = p.Id,
    TicketName = p.Ticket.Name,
    CustomFields = p.CustomFieldValues.Select(cf => new
    {
        Name = cf.CustomFieldDefinition.Name,
        Value = cf.Value
    }).ToList()
}).ToLinqToDB().ToList();
```

This guery yields back 300 results, from the logging it seems that Linq2Db core runs the subquery for each of the obtained item. [= 300 subqueries, 300 roundtrips to the DB]. For our scenario, for example, this creates an issue as our SQL Server is physically located elsewhere than the calling app creating a huge performance gap due to the latency, but I believe this might be an issue for other scenarios too (running 301 queries when in fact only 2 are needed)

From observing the logs EF Core handles this differently - running a single subquery on the very same condition as main "root" query, effectively creating only one more roundtrip to the DB, then linking the data based on their foreign-key declaration.

To sum it up, I have 3 questions:
- Is there anything I can do to change the one-to-many behavior?
- Are there any plans to change the subdata loading routine?
- If I wanted to hack a solution for our needs by changing the Linq2Db core, could you advise me where should I look?

**Comment by MaceWindu** [comment]
Yes, unfortunatelly, this is how it currently works. We have several related issues filled for it already https://github.com/linq2db/linq2db/issues?q=is%3Aissue+is%3Aopen+label%3A%22epic%3A+eager-load%22.

As workaround I would recommend to split your query into two queries and manually set CustomFileds value using data from second query for now.

Don't think it will be an easy task to hack something into linq2db, but if you want to try, maybe @sdanyliv could provide some tips

**Comment by brunolau** [comment]
Unfortunately truth on that, I gave up. From looking at the source I believe it would take a tremendous effort to implement N2M in a clean way. I think I had a lead on how to trick it thou in a very dirty way (I even managed to get some simple scenarios running, but as I'm nowhere near experienced with writing expressions I gave up.) I've got lost in not being able to meaningfully pass the parent collection of TicketNumbers into the subquery cycle. Anyway, I'm sharing my concept, I've attempted to create on synchronous ToList method creating from projection (I've noticed that LoadWith method is handled differently)... someone perhaps might pick up on that. 

My idea - run the parent query, then make the child collection load by injecting relevant foreign keys into one single query [e.g. SELECT * FROM CustomFieldValue WHERE TicketNumberId IN (1,2,3...) ...might be not optimal for bigger results list but still performs better than loading one-by-one.

Started by injecting ExpressionBuilder.QueryBuilder.GetMultipleQueryExpression to contain a list

```
for (var i = 0; i < args.Association.ThisKey.Length; i++)
{
	var field1 = args.ParentSqlTable.Fields[args.Association.ThisKey[i]];
	var field2 = args.SqlTable.Fields[args.Association.OtherKey[i]];
	var exChild = Expression.MakeMemberAccess(op, field2.ColumnDescriptor.MemberInfo);
	var exParent = Expression.MakeMemberAccess(me.Expression, field1.ColumnDescriptor.MemberInfo);

	//Add list-type parameter comparison, creates WHERE x IN (...y)
	var listType = typeof(IdList<>);
	var constructedListType = listType.MakeGenericType(field1.ColumnDescriptor.MemberType);
	var containsMethod = listType.GetMethod("Contains");
	var listInstance = Activator.CreateInstance(constructedListType) as IIdList;
	listInstance.ParentName = field1.ColumnDescriptor.MemberName;
	listInstance.ParentType = field1.ColumnDescriptor.MemberType;
	listInstance.ChildName = field2.ColumnDescriptor.MemberName;
	listInstance.ChildType = field2.ColumnDescriptor.MemberType;
	subQueryDeclaration.PropertyLists.Add(listInstance);

	var listInstanceExpr = Expression.Constant(listInstance);
	var expContains = Expression.Call(typeof(Enumerable), "Contains", new[] { exChild.Type }, listInstanceExpr, exChild);
	var dummyExpParent = Expression.Equal(exParent, exParent); //Trick to ensure Key is in the parent select. This will create something like 1=1 on the child
	var ee = Expression.AndAlso(expContains, dummyExpParent);
	ex = ex == null ? ee : Expression.AndAlso(ex, ee);
}
```

In Linq.QueryRunner.ExecuteQuery I've first read all the results and then started yielding.

```
var resultList = new DataReaderResultCollection()
using (var dr = runner.ExecuteReader())
{
	while (dr.Read())
	{
		var resultRow = new DataReaderResultRow() { ParentCollection = resultList };
		for (int i = 0; i < dr.FieldCount; i++)
		{
			resultRow.AddItem(i, new DataReaderResultRowItem
			{
				DataTypeName = dr.GetDataTypeName(i),
				FieldType = dr.GetFieldType(i),
				Name = dr.GetName(i),
				Value = dr.GetValue(i)
			});
		}

		resultList.Add(resultRow);
	}
}

foreach (var drRow in resultList)
{
	var value = mapper.Map(runner, drRow);
	runner.RowsCount++;
	yield return value;
}
```

Here comes the tricky part. As the library stack seems to rely heavily on DataReader, I've attempted to overcome the issue by implementing a virtual datareader. The result row class DataReaderResultRow inherits from DbDataReader allowing rest of the uninjected stack work as before.

```
public class DataReaderResultRow : DbDataReader
{
	private SortedList<int, DataReaderResultRowItem> _list = new SortedList<int, DataReaderResultRowItem>();
	internal DataReaderResultCollection ParentCollection { get; set; }

	public override int FieldCount => _list.Count;
	public override bool HasRows => _list.Count > 0;

	public override object this[string name] => throw new NotImplementedException();
	public override object this[int i] => _list[i];

	internal void AddItem(int index, DataReaderResultRowItem item)
	{
		_list.Add(index, item);
	}
	
	public override string GetName(int i)
	{
		return _list[i].Name;
	}

	public override string GetDataTypeName(int i)
	{
		return _list[i].DataTypeName;
	}

	public override Type GetFieldType(int i)
	{
		return _list[i].FieldType;
	}

	public override object GetValue(int i)
	{
		return _list[i].Value;
	}

	public override int GetOrdinal(string name)
	{
		for (int i = 0; i < _list.Count; i++)
		{
			var item = _list[i];
			if (item.Name == name)
			{
				return i;
			}
		}

		return -1;
	}

	public override bool GetBoolean(int i)
	{
		return Convert.ToBoolean(_list[i].Value);
	}

	public override byte GetByte(int i)
	{
		return Convert.ToByte(_list[i].Value);
	}

	public override char GetChar(int i)
	{
		return Convert.ToChar(_list[i].Value);
	}

	public override Guid GetGuid(int i)
	{
		return new Guid(_list[i].Value.ToString());
	}

	public override short GetInt16(int i)
	{
		return Convert.ToInt16(_list[i].Value);
	}

	public override int GetInt32(int i)
	{
		return Convert.ToInt32(_list[i].Value);
	}

	public override long GetInt64(int i)
	{
		return Convert.ToInt64(_list[i].Value);
	}

	public override float GetFloat(int i)
	{
		return Convert.ToSingle(_list[i].Value);
	}

	public override double GetDouble(int i)
	{
		return Convert.ToDouble(_list[i].Value);
	}

	public override string GetString(int i)
	{
		return _list[i].Value.ToString();
	}

	public override decimal GetDecimal(int i)
	{
		return Convert.ToDecimal(_list[i].Value);
	}

	public override DateTime GetDateTime(int i)
	{
		return (DateTime)_list[i].Value;
	}

	public override bool IsDBNull(int i)
	{
		return (_list[i].Value == DBNull.Value || _list[i].Value == null);
	}

	public override int Depth => throw new NotImplementedException();
	public override bool IsClosed => throw new NotImplementedException();
	public override int RecordsAffected => throw new NotImplementedException();

	public override int GetValues(object[] values)
	{
		throw new NotImplementedException();
	}

	public override bool NextResult()
	{
		throw new NotImplementedException();
	}

	public override bool Read()
	{
		throw new NotImplementedException();
	}

	public override long GetBytes(int i, long fieldOffset, byte[] buffer, int bufferoffset, int length)
	{
		throw new NotImplementedException();
	}

	public override long GetChars(int i, long fieldoffset, char[] buffer, int bufferoffset, int length)
	{
		throw new NotImplementedException();
	}

	public override IEnumerator GetEnumerator()
	{
		throw new NotImplementedException();
	}

	public override DataTable GetSchemaTable()
	{
		return ParentCollection.SchemaTable;
	}
}
```

Up until now kinda simple, but this is where I got lost. I couldn't find a way on how to pass parent data collection to children. I've tried some black-magic solutions such playing with static data [assigning some additional contextual data to the DataContext class]. That did kinda work and if converted to WeakReference and disposed properly might even mitigate the risk of memory leak, but I've run into problems with query caching and others, so basically without knowing how to effectively share these contexts between parent/children contexts I'm pointless.

I'm attaching the ZIP of my attempt, most of the code is in "ConsoleApplication1", I've also had to slightly change some methods in the library core in order to allow injection/customization
[linq2db-master.zip](https://github.com/linq2db/linq2db.EntityFrameworkCore/files/3051983/linq2db-master.zip)

**Comment by brunolau** [comment]
To add a bit. If someone could help me with how to pass the parent/children context I might look into this as this. I've tried to make my approach optional with as litle as possible intereference with the core. If you look at the ZIP, I believe I've added no maintenance or performance overhead to existing state so if pointed to the right direction, this could live it's life as some kind of extension library outside of the l2db core for example.

**Comment by MaceWindu** [comment]
You need to wait for @sdanyliv (he should be back in several days) - maybe he will have some ideas how to help you

**Comment by sdanyliv** [comment]
@brunolau, Eager  Loading in our TODO list a long time. I'm pretty sure that it is not possible to implement it as separate library. I'll take a look again at this task, and if it is doable with current linq parser - i'll try to implement this one. 

BTW, i'm working on new experimental linq parser and it should contain this functionality, but i do not recommend to wait this in nearest time. A lot of work here.

Also moving away from EF Core to linq2db can be challenging task, because we do not have ChangeTracker, so you have to review your data modification parts. ChangeTracker can be useful in many scenarios, but from my experience - you can always avoid its usage.

**Comment by brunolau** [comment]
@sdanyliv That's great news, I'll be waiting impatiently :) ... about moving away from EF Core I more thought about using EF Core+LINQ2DB integration exclusively..now our code is full of extension methods such as ToListLinq2Db, ToListEfCore and similar, because although most queries are handled by LINQ2DB way better, because of the eager loading I still have to use EfCore somewhere...so hopefully one day we'll be able to use LINQ2DB exclusively. 

Any rough (rounded to years) ETA on the experimental parser?

**Comment by nathan-parkinson** [comment]
I have been working on a library to add eager loading to linq2db that might help with your issue until such a feature is implemented in linq2db itself, it's called [linq2db.include](https://github.com/nathan-parkinson/linq2db.include). 

I hope it is ok to mention it here, I don't want to tread on anyone's toes.  

It is still in alpha but this is only because I am yet to implement automatic matching of back-references and need to rework the custom overrides that allow developers to manually change the way child data is loaded.  If you do not require these features then it is safe to use. We are using it in production at my place of work without issues.

Also @sdanyliv's comment
> I'm pretty sure that it is not possible to implement it as separate library.

is not lost on me.  I have not had any issues but am now wondering what unknown issues I may be about to experience.  Any more explanation from @sdanyliv would be greatly appreciated.

**Comment by brunolau** [comment]
@nathan-parkinson does this work also when using projection, or only replaces .Include() functionality ?

**Comment by nathan-parkinson** [comment]
Sorry for the late reply @brunolau.  

No, it doesn't do projection.  I think that would have to be more of an integrated feature.  
The purpose of this project was only to replicate the .Include() functionality.

**Comment by sdanyliv** [comment]
@brunolau, good news there. I had created first Eager Loading prototype using current sources and It looks very promising, later will point to PR which introduces this feature. A lot of work here but i hope in one or two weeks it will be done.

**Comment by brunolau** [comment]
@sdanyliv That's some great news, will be waiting for the PR

**Comment by sdanyliv** [comment]
PR, which introduces Eager Loading is here https://github.com/linq2db/linq2db/pull/1756
Will try to cover as many cases as it is possible.

**Comment by sdanyliv** [comment]
It was long journey... and it was not easy. But finally we have implemented that, and I think it is much better than in EF Core 2.1 and non-comapriable to EF Core 3.1, they turned back to EF 6.X implementation with cartesian explosion issues.

Expect this feature in upcoming `3.0.0-preview.2` release.

**Comment by brunolau** [comment]
Any ETA on the preview.2 release date ?

**Comment by sdanyliv** [comment]
In hope tomorrow.

**Comment by sdanyliv** [comment]
Released in `3.0.0-preview.2`
https://github.com/linq2db/linq2db/wiki/Release-Notes-3.0.0#preview-2-changes
